xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -624,21 +624,39 @@ private void handleRdsResponse(DiscoveryResponse rdsResponse) {
   }
 
   /**
-   * Processes RouteConfiguration message (from an resource information in an LDS or RDS
-   * response), which may contain a VirtualHost with domains matching the "xds:"
-   * URI hostname directly in-line. Returns the clusterName found in that VirtualHost
-   * message. Returns {@code null} if such a clusterName cannot be resolved.
+   * Processes a RouteConfiguration message to find the name of upstream cluster that requests
+   * for the given host will be routed to. Returns the clusterName if found.
+   * Otherwise, returns {@code null}.
    *
-   * <p>Note we only validate VirtualHosts with domains matching the "xds:" URI hostname.
    */
+  @VisibleForTesting
   @Nullable
-  private String processRouteConfig(RouteConfiguration config) {
+  static String findClusterNameInRouteConfig(RouteConfiguration config, String hostname) {
     List<VirtualHost> virtualHosts = config.getVirtualHostsList();
-    int matchingLen = -1;  // longest length of wildcard pattern that matches host name
+    // Domain search order:
+    //  1. Exact domain names: ``www.foo.com``.
+    //  2. Suffix domain wildcards: ``*.foo.com`` or ``*-bar.foo.com``.
+    //  3. Prefix domain wildcards: ``foo.*`` or ``foo-*``.
+    //  4. Special wildcard ``*`` matching any domain.
+    //
+    //  The longest wildcards match first.
+    //  Assuming only a single virtual host in the entire route configuration can match
+    //  on ``*`` and a domain must be unique across all virtual hosts.
+    int matchingLen = -1; // longest length of wildcard pattern that matches host name
     VirtualHost targetVirtualHost = null;  // target VirtualHost with longest matched domain
     for (VirtualHost vHost : virtualHosts) {
       for (String domain : vHost.getDomainsList()) {
-        if (matchHostName(hostName, domain) && domain.length() > matchingLen) {
+        boolean selected = false;
+        if (matchHostName(hostname, domain)) { // matching
+          if (!domain.contains("*")) { // exact matching
+            selected = true;
+          } else if (domain.length() > matchingLen) { // longer matching pattern
+            selected = true;
+          } else if (domain.length() == matchingLen && domain.startsWith("*")) { // suffix matching
```

Comment:
If both `a.example.com` (exact matching) and `*.example.com` (suffix matching ) are in the list, then suffix matching will be finally selected if it comes later.

Reply:
Yep. Fix.